lib/segment/src/cbor_serde.rs

```diff
@@ -0,0 +1,81 @@
+//! This module encapsulate all the CBOR serde logic.
+
+use std::io::{Read, Write};
+use std::vec::Vec;
+
+use serde::{de, ser};
+
+#[derive(Debug)]
+pub struct Error {
+    pub description: String,
+}
+
+impl From<ciborium::ser::Error<std::io::Error>> for Error {
+    fn from(err: ciborium::ser::Error<std::io::Error>) -> Self {
+        Error {
+            description: format!("CBOR serde error: {:?}", err),
+        }
+    }
+}
+
+impl From<ciborium::de::Error<std::io::Error>> for Error {
+    fn from(err: ciborium::de::Error<std::io::Error>) -> Self {
+        Error {
+            description: format!("CBOR serde error: {:?}", err),
+        }
+    }
+}
+
+/// Serializes as CBOR into a [Writer](std::io::Write)
+#[inline]
+pub fn into_writer<T: ?Sized + ser::Serialize, W: Write>(
+    value: &T,
+    writer: W,
+) -> Result<(), Error> {
+    ciborium::ser::into_writer(value, writer)?;
+    Ok(())
+}
+
+/// Deserializes as CBOR from a [Reader](std::io::Read)
+#[inline]
+pub fn from_reader<'de, T: de::Deserialize<'de>, R: Read>(reader: R) -> Result<T, Error> {
+    let result = ciborium::de::from_reader(reader)?;
+    Ok(result)
+}
+
+/// Serializes as CBOR into a [Vec](std::vec::Vec)
+pub fn into_vec<T: ?Sized + ser::Serialize>(value: &T) -> Result<Vec<u8>, Error> {
```

Comment:
why not `to_vec`? it doesn't consume the `value`, right

Reply:
see comment above ☝️